### PR TITLE
Support version selection for database plugins

### PR DESCRIPTION
**Pull request by tomhjp**
Supports specifying a `plugin_version` when configuring a database plugin.

To test locally:

```bash
make dev
go build -o plugin_directory/postgres-1.11.0-rc1 plugins/database/postgresql/postgresql-database-plugin/main.go
cp plugin_directory/postgres-1.11.0-rc1 plugin_directory/postgres-1.11.0-rc2

# Run postgres + vault
docker run --rm --name postgres -e POSTGRES_USER=root -e POSTGRES_PASSWORD=root -d -p 5432:5432 postgres
VAULT_DEV_ROOT_TOKEN_ID=root vault server -dev -dev-plugin-dir=pluin_directory &
export VAULT_ADDR=http://127.0.0.1:8200
export VAULT_TOKEN=root

# Register and configure plugins
SHA256="$(shasum -a 256 plugin_directory/postgres-1.11.0-rc1 | awk '{print $1}')"
vault write sys/plugins/catalog/database/pg sha256=$SHA256 command=postgres-1.11.0-rc1 version=v1.11.0-rc1
vault write sys/plugins/catalog/database/pg sha256=$SHA256 command=postgres-1.11.0-rc2 version=v1.11.0-rc2
vault secrets enable database

# Explicitly use rc1
vault write database/config/pg1 plugin_name=pg connection_url="postgresql://{{username}}:{{password}}@localhost:5432/?sslmode=disable" allowed_roles="*" username=root password=root plugin_version=1.11.0-rc1

# This will select and pin the latest versioned pg plugin registered, rc2
vault write database/config/pg2 plugin_name=pg connection_url="postgresql://{{username}}:{{password}}@localhost:5432/?sslmode=disable" allowed_roles="*" username=root password=root

# Prove the correct binaries are running
ps aux | grep postgres
```

### CODE REVIEW

File: vault/mount.go [hunk 1]
```diff
@@ -306,7 +306,7 @@ const mountStateUnmounting = "unmounting"
 type MountEntry struct {
 	Table                 string            `json:"table"`                             // The table it belongs to
 	Path                  string            `json:"path"`                              // Mount Path
-	Type                  string            `json:"type"`                              // Logical backend Type
+	Type                  string            `json:"type"`                              // Logical backend Type. NB: This is the plugin name, e.g. my-vault-plugin, NOT plugin type (e.g. auth).
```
Comment (mpalmi) [inline vault/mount.go:309]: +1 this is a confusing overloading of terms. A bit of clarity is good.

File: sdk/logical/system_view.go [hunk 2]
```diff
@@ -176,6 +180,10 @@ func (d StaticSystemView) LookupPluginVersion(_ context.Context, _ string, _ con
 	return nil, errors.New("LookupPluginVersion is not implemented in StaticSystemView")
 }
 
+func (d StaticSystemView) ListVersionedPlugins(_ context.Context, _ consts.PluginType) ([]pluginutil.VersionedPlugin, error) {
+	return nil, errors.New("ListVersionedPlugins is not implemented in StaticSystemView")
```
Comment (mpalmi) [inline sdk/logical/system_view.go:184]: Is there a reason (besides matching surrounding style) to use `errors.New()` here and `fmt.Errorf()` in other places? Pretty minor detail, but just something that occurred to me when comparing this to https://github.com/hashicorp/vault/pull/16982/files#diff-321fc062fce4a794ac87052c0863ebd4177ac8e21d9cb1b718bf21c1851b5febR264

**Review by mpalmi — APPROVED** [review]
Looks good to me. I haven't had the chance to take it for a spin yet, but that shouldn't hold up the merge. Nor should the nits in my review.

**Review by swenson — COMMENTED** [review]
So far, looks great.

File: sdk/logical/system_view.go (hunk 2 above)
Comment (tomhjp) [reply sdk/logical/system_view.go:184]: My personal preference is for `errors.New` when there are no format arguments, but functionally they're the same, and linters will catch mismatches so I just chose to match the local style.

**Comment by tomhjp** [comment]
Thanks for the reviews so far - I think it's ready for another pass @swenson

**Review by swenson — APPROVED** [review]
LGTM

**Pull request merged**